src/zkp/ecdsa_adaptor.rs

```diff
@@ -0,0 +1,206 @@
+    pub fn adaptor_decrypt(&self, decryption_key: &SecretKey) -> Signature {
+        unsafe {
+            let mut signature = ffi::Signature::new();
+            assert_eq!(
+                1,
+                ffi::secp256k1_ecdsa_adaptor_decrypt(
+                    ffi::secp256k1_context_no_precomp,
+                    &mut signature,
+                    decryption_key.as_c_ptr(),
+                    self.as_c_ptr(),
+                )
+            );
+            Signature::from(signature)
+        }
+    }
+
+    /// Extracts the adaptor secret from the complete signature and the adaptor signature.
+    pub fn adaptor_recover<C: Signing>(
+        &self,
+        secp: &Secp256k1<C>,
+        sig: &Signature,
+        encryption_key: &PublicKey,
+    ) -> Result<SecretKey, Error> {
+        let mut data: [u8; constants::SECRET_KEY_SIZE] = [0; constants::SECRET_KEY_SIZE];
+
+        unsafe {
+            assert_eq!(
+                1,
+                ffi::secp256k1_ecdsa_adaptor_recover(
+                    *secp.ctx(),
+                    data.as_mut_c_ptr(),
+                    sig.as_c_ptr(),
+                    self.as_c_ptr(),
+                    encryption_key.as_c_ptr()
+                )
+            );
+        }
+
+        Ok(SecretKey::from_slice(&data)?)
+    }
+
+    /// Verifies that the adaptor secret can be extracted from the adaptor signature and the completed ECDSA signature.
+    pub fn adaptor_verify<C: Verification>(
+        &self,
+        secp: &Secp256k1<C>,
+        msg: &Message,
+        pubkey: &PublicKey,
+        encryption_key: &PublicKey,
+    ) -> Result<(), Error> {
+        unsafe {
+            let res = ffi::secp256k1_ecdsa_adaptor_verify(
+                *secp.ctx(),
+                self.as_c_ptr(),
+                pubkey.as_c_ptr(),
+                msg.as_c_ptr(),
+                encryption_key.as_c_ptr(),
+            );
+            return if res == 1 {
+                Ok(())
+            } else {
+                Err(Error::Upstream(UpstreamError::IncorrectSignature))
+            };
+        }
+    }
+}
+
+#[cfg(all(test, feature = "global-context"))]
+mod tests {
+    use super::*;
+    use super::{Message, Secp256k1};
+    use rand::thread_rng;
+
+    #[test]
+    fn test_ecdsa_adaptor_signature() {
+        let secp = Secp256k1::new();
+
+        let (seckey, pubkey) = secp.generate_keypair(&mut thread_rng());
+        let (adaptor_secret, adaptor) = secp.generate_keypair(&mut thread_rng());
+        let msg = Message::from_slice(&[2u8; 32]).unwrap();
+        let adaptor_sig = EcdsaAdaptorSignature::adaptor_encrypt(&secp, &msg, &seckey, &adaptor);
+
+        assert!(adaptor_sig
+            .adaptor_verify(&secp, &msg, &pubkey, &adaptor)
+            .is_ok());
+        assert!(adaptor_sig
+            .adaptor_verify(&secp, &msg, &adaptor, &pubkey)
+            .is_err());
+        let sig = adaptor_sig.adaptor_decrypt(&adaptor_secret);
+        #[cfg(not(rust_secp_fuzz))]
```

Comment:
This doesn't feel super clean, but I don't have a very good idea on how to implement a version of `adaptor_decrypt` that would be compatible with the fuzz version of `verify` since it doesn't have access to the `msg`.

Reply:
I'd rather have us disable to whole test to be honest. I am not sure we need to run them as part of fuzzing.

Reply:
Done, sounds good to me.